### ISSUE TITLE
activePollIntervalMs appears to be broken

### ISSUE THREAD

**Issue by upugo-dev** [PruvoNet/squiss-ts#30]
As far as I understand it, activePollIntervalMs should delay the poller by a specified number of ms before picking up another message? If so, I don't seem to be able to get it to work. As soon as one message is processed, the next message is picked up with no delay, even though I have set:

`activePollIntervalMs: 5000,`

Any ideas?

**Comment by regevbr** [comment]
The library is intended to always keep getting messages into the system for full utilization, but never more than the desired handle count.

Basically the way the library works is as follows:
- As long as there are more message slots available to handle
  - Query SQS for new messages (waits for `receiveWaitTimeSecs` time until it aborts the query) 
    - If received message(s) and there are still slots available, wait `activePollIntervalMs` and start over
    - If received no messages and there are still slots available, wait `idlePollIntervalMs` and start over
    - If there are no more slots available, pause and wait for a slot to be available (e.g. a message is marked as handled)
      - Once a slot is available (e.g a message was handled), start from the beginning **immediately**

So basically:
- the `activePollIntervalMs` option is used to not choke the SQS api, while actively waiting for message for messages after some were received
- the `idlePollIntervalMs` option is used to not choke the SQS api, while actively waiting for message for messages after no messages were received

And currently there is no option to delay after a slot becomes available.

If you want, I can add it, but I really don't see the point in that. The reason the 2 options above are set to `0` by default is to achieve full utilization and always handle as much messages as possible...

Thoughts?

**Comment by upugo-dev** [comment]
Ah ok, I understand, thanks for your explanation! No, don't add a feature you think is pointless on my behalf! 

My use case actually requires that I process messages strictly one-by-one, so I have set maxInFlight to 1 anyway. 

I was hoping that I could use the messaging service to somehow delay message consumption, allowing the processor a 'cooldown' period - this is to allow the processor a few seconds to properly close some chromium processes and so on. Basically to become ready to receive the next message properly. I will write this logic myself though, it will be better.

As an aside, this application will be deployed and scaled horizontally, meaning there will be a single sqs-queue with many pollers. (1 poller per node.js application, each on its own amazon ec2 instance). Will setting maxInFlight = 1 affect the behaviour of multiple pollers on a single queue?

**Comment by regevbr** [comment]
That is indeed a valid use case, but as you said, it sounds like you should not mark the message as handled, until you are ready to handle the next one - because even if I will provide such options, it it just an arbitrary time, wheres you should control that time according to your need and they can vary between different messages.

This library is intended exactly for the scaling you described and the proper usage for it!

**Comment by upugo-dev** [comment]
OK, perfect, I will do as you suggest.

However, what about when the message times out? Then, there is not an opportunity to let the processor settle before marking the message as handled, because the message is marked as handled automatically by the timeout event before the 'timeoutReached' is fired, and so any processor-cooldown will be ongoing while/after the processor has picked up the next message

**Comment by regevbr** [comment]
I agree that there might be some issue there, but what do you suggest? or how would you want it to be?

**Comment by regevbr** [comment]
To tell you the truth, I don't use SQS as a singleton runner like you, so your use cases are quite different. I'm willing (and happy!) to make all the modification to support your use cases, but I can't decide with myself what should be the appropriate flow...

The reason I use the timeout extender, is to prevent message starvation and under utilization:
You can skip the timeout extender mechanism entirely, just by setting the message visibility timeout to a very large number, say 12 hours, and then just ack/nack it when you want. Things start to go sideways if you fail to ack/nack it in time (if the process crashes for example) - this will cause the message to get resubmitted after 12 hours! which is insane.

To deal with the bad scenario, I set ( depending on the queue characteristics) a low visibility timeout (usually 5 minutes) and then tell the library to renew it a few times after which I consider the message as stale and because I have no way to actually stop it from being processed (maybe you do...) I just continue on to the next message.

**Comment by upugo-dev** [comment]
> I agree that there might be some issue there, but what do you suggest? or how would you want it to be? 

I guess there is no solution from the sqs side, the logic that I want should be contained in my application 😄 

> The reason I use the timeout extender, is to prevent message starvation and under utilization:
Yes this is the same reason as us; ensure that messages are being locked for the minimum amount of time to try and maximize available resource. If we didn't care about resource, then this wouldn't be an issue.

Interesting idea with the "long message-ack/nack when done" approach, I may resort to that eventually, but since I am so close to a solution with the extension approach, I am inclined to continue. Waiting 12h for the message to come round again would not be acceptable in our production environment.

> To deal with the bad scenario, I set ( depending on the queue characteristics) a low visibility timeout (usually 5 minutes) and then tell the library to renew it a few times after which I consider the message as stale and because I have no way to actually stop it from being processed (maybe you do...) I just continue on to the next message.

Yeah, this is effectively my exact solution, except at the time when the message has fully timed out, we cancel any ongoing processes. The issue I've been running in to, is the fact that the cancellation of the processes can take longer than the time taken to receive the next message, which interferes with processing the new message. I'm thinking of solving this by forking the node.js process so that each 'message process' is a separate process and so cannot interfere.

**Comment by regevbr** [comment]
Ok that makes sense!
If there is nothing you think I can help with here on my side, please close the ticket.

And again, I will be happy to continue to assist you in getting your requirements met.

**Comment by upugo-dev** [comment]
thanks man, you are awesome
